File: apps/multimodal-screener/iOS/Aiberry/Emotion/EmotionModel.swift
import Foundation
import ZeticMLange

/// Speech-emotion recognition on-device via `realtonypark/Wav2Vec2-Base_Emotion-Recognition`
/// (wav2vec2-large-xlsr, Apache-2.0). Input: raw 16 kHz waveform `(1, 48000)`; output: 7 logits.
///
/// This is the **voice** half of the Aiberry-style multimodal check-in — the same
/// pipeline a client would use, just with their model. Reused verbatim from VoiceVitals.
final class EmotionModel: ObservableObject {
    @Published var status: ModelStatus = .idle
    @Published var scores: [EmotionScore] = []
    @Published var latencyMs: Double?

    private var model: ZeticMLangeModel?
    private let queue = DispatchQueue(label: "aiberry.emotion")

    // Label order matches the model's config.id2label:
    // [angry, disgust, fear, happy, neutral, sad, surprise].
    let labels = ["Angry", "Disgust", "Fear", "Happy", "Neutral", "Sad", "Surprise"]

    var top: EmotionScore? { scores.max(by: { $0.probability < $1.probability }) }

    /// Download + compile the model ahead of time (called at launch).
    func preload() {
        guard model == nil, !status.isBusy else { return }
        status = .downloading(0)
        queue.async { [weak self] in
            guard let self else { return }
            do { _ = try self.ensureModel(); DispatchQueue.main.async { self.status = .idle } }
            catch { DispatchQueue.main.async { self.status = .failed(String(describing: error)) } }
        }
    }

    /// Analyze a captured clip and publish ranked scores. `completion` returns the
    /// final probability distribution (label order = `labels`) for the fusion engine.
    func analyze(_ samples: [Float], completion: (([Float]) -> Void)? = nil) {
        guard !status.isBusy else { return }
        status = .running

        queue.async { [weak self] in
            guard let self else { return }
            do {
                let model = try self.ensureModel()
                let n = AppConfig.clipSamples

                // 1) Trim dead air so the model's mean-pool sees mostly speech.
                let speech = AudioUtils.trimSilence(samples)
                // 2) Cover the whole utterance: average logits over overlapping windows.
                let windows = AudioUtils.windows(speech, size: n)
                    .map { MelangeKit.fitTiling($0, to: n) }

                var summed = [Float](repeating: 0, count: self.labels.count)
                var totalMs = 0.0
                for window in windows {
                    let input = MelangeKit.floatTensor(window, shape: [1, n]) // (1, 48000)
                    let (outputs, ms) = try measureMs { try model.run(inputs: [input]) }
                    totalMs += ms
                    guard let logits = outputs.first.map({ MelangeKit.floats(from: $0) }),
                          logits.count >= self.labels.count else {
                        throw SimpleError("Unexpected model output (\(outputs.first?.shape ?? []))")
                    }
                    for i in 0..<self.labels.count { summed[i] += logits[i] }
                }
                let logits = summed.map { $0 / Float(windows.count) }
                // Temperature scaling (T>1) softens over-confident logits; display only.
                let temperature: Float = 2.0
                let scaled = logits.prefix(self.labels.count).map { $0 / temperature }
                let probs = MelangeKit.softmax(Array(scaled))
                let ranked = zip(self.labels, probs)
                    .map { EmotionScore(label: $0.0, probability: $0.1) }
                    .sorted { $0.probability > $1.probability }
                DispatchQueue.main.async {
                    self.scores = ranked
                    self.latencyMs = totalMs
                    self.status = .ready
                    completion?(probs)
                }
            } catch {
                DispatchQueue.main.async {
                    self.status = .failed(String(describing: error))
                    completion?([])
                }
            }
        }
    }

    private func ensureModel() throws -> ZeticMLangeModel {
        if let model { return model }
        let loaded = try MelangeKit.load(AppConfig.Model.emotion, version: AppConfig.Model.emotionVersion) { progress in
            DispatchQueue.main.async { self.status = .downloading(progress) }
        }
        self.model = loaded
        return loaded
    }
}

File: apps/multimodal-screener/Android/app/src/main/java/ai/zetic/aiberry/emotion/EmotionModel.kt
package ai.zetic.aiberry.emotion

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import ai.zetic.aiberry.core.AppConfig
import ai.zetic.aiberry.core.AudioUtils
import ai.zetic.aiberry.core.MelangeKit
import ai.zetic.aiberry.core.MelangeRuntime
import ai.zetic.aiberry.core.ModelStatus
import ai.zetic.aiberry.core.measureMs
import com.zeticai.mlange.core.model.ZeticMLangeModel

data class EmotionScore(val label: String, val probability: Float)

/**
 * Speech-emotion recognition on-device via `realtonypark/Wav2Vec2-Base_Emotion-Recognition`
 * (wav2vec2-base, Apache-2.0). Input: raw 16 kHz waveform `(1, 48000)`; output: 7 logits.
 *
 * The cleanest stand-in for a mental-health vocal biomarker: the same pipeline a
 * client would use, just with their model.
 */
class EmotionModel(private val context: Context) {
    var status by mutableStateOf<ModelStatus>(ModelStatus.Idle)
        private set
    var scores by mutableStateOf<List<EmotionScore>>(emptyList())
        private set
    var latencyMs by mutableStateOf<Double?>(null)
        private set

    private var model: ZeticMLangeModel? = null
    // Shared across all models so ZeticMLange init/run never overlap (see MelangeRuntime).
    private val queue = MelangeRuntime.executor
    private val main = Handler(Looper.getMainLooper())

    // Label order matches the model's config.id2label:
    // [angry, disgust, fear, happy, neutral, sad, surprise].
    private val labels = listOf("Angry", "Disgust", "Fear", "Happy", "Neutral", "Sad", "Surprise")

    val top: EmotionScore? get() = scores.maxByOrNull { it.probability }

    /** Download + compile the model ahead of time (called at launch). */
    fun preload() {
        if (model != null || status.isBusy) return
        status = ModelStatus.Downloading(0f)
        queue.execute {
            try {
                ensureModel()
                main.post { status = ModelStatus.Idle }
            } catch (e: Throwable) {
                main.post { status = ModelStatus.Failed(MelangeKit.friendly(e)) }
            }
        }
    }

    /**
     * Analyze a waveform. [onResult], if given, receives the 7 softmax probabilities in
     * canonical [emotionLabels] order on the main thread — this is what the multimodal
     * fusion engine consumes (matches iOS `voice.analyze(concat) { probs in … }`).
     */
    fun analyze(samples: FloatArray, onResult: ((FloatArray) -> Unit)? = null) {
        if (status.isBusy) {
            onResult?.let { cb -> main.post { cb(FloatArray(0)) } }
            return
        }
        status = ModelStatus.Running

        queue.execute {
            try {
                val model = ensureModel()
                val n = AppConfig.clipSamples

                // 1) Trim dead air so the model's mean-pool sees mostly speech.
                val speech = AudioUtils.trimSilence(samples)
                // 2) Cover the whole utterance: average logits over overlapping windows
                //    (instead of cropping to the middle 3 s). Each window is tiled to 48000.
                val windows = AudioUtils.windows(speech, n).map { MelangeKit.fitTiling(it, n) }

                val summed = FloatArray(labels.size)
                var totalMs = 0.0
                for (window in windows) {
                    val input = MelangeKit.floatTensor(window, intArrayOf(1, n)) // (1, 48000)
                    val (outputs, ms) = measureMs { model.run(arrayOf(input)) }
                    totalMs += ms
                    val logits = outputs.firstOrNull()?.let { MelangeKit.floats(it) }
                    if (logits == null || logits.size < labels.size) {
                        throw IllegalStateException("Unexpected model output")
                    }
                    for (i in labels.indices) summed[i] += logits[i]
                }
                val logits = FloatArray(labels.size) { summed[it] / windows.size }

                // Temperature scaling (T>1) softens the model's over-confident, near
                // one-hot logits so secondary emotions are visible. Same winner; display only.
                val temperature = 2.0f
                val scaled = FloatArray(labels.size) { logits[it] / temperature }
                val probs = MelangeKit.softmax(scaled)
                val ranked = labels.indices
                    .map { EmotionScore(labels[it], probs[it]) }
                    .sortedByDescending { it.probability }

                main.post {
                    scores = ranked
                    latencyMs = totalMs
                    status = ModelStatus.Ready
                    onResult?.invoke(probs)
                }
            } catch (e: Throwable) {
                main.post {
                    status = ModelStatus.Failed(MelangeKit.friendly(e))
                    onResult?.invoke(FloatArray(0))
                }
            }
        }
    }

    @Synchronized
    private fun ensureModel(): ZeticMLangeModel {
        model?.let { return it }
        val loaded = MelangeKit.load(
            context,
            AppConfig.Model.EMOTION,
            AppConfig.Model.EMOTION_VERSION,
        ) { progress -> main.post { status = ModelStatus.Downloading(progress) } }
        model = loaded
        return loaded
    }
}
